### ISSUE TITLE
order of events -- where defined

### ISSUE THREAD

**Issue by johanneswilm** [w3c/input-events#22]
I cannot see that we came up with a solution where the order of events for these should be defined:

beforeinput/input events in relation to:
- clipboard: paste/cut
- history: undo/redo
- dragging: drag/drop

**Comment by garykac** [comment]
The proposal was for a separate document that contained the canonical list of events in order.

The reasoning was that:
(1) the core spec would only cover the core events and would not have info on any new events
(2) specs that define new events can't reasonably include info about events in every other spec. Esp, if we consider that many specs will be in-flight simultaneously in various stages of completeness.

The thought was to make this a very lightweight "spec" that contained only the ordering information (and perhaps examples?).

This is similar to the approach used for the UI Events code and key specs, which are basically just tables of key codes. This makes them easy to update without opening up the entire core UI Events spec.

**Comment by johanneswilm** [comment]
@garykac OK, I will add a simply "light" spec for this.

**Comment by annevk** [comment]
I don't think that works. We need to actually define the action, corresponding task, and what happens with each event dispatched from that task. How canceling affects any subsequent events, etc. Just putting event names in a list won't do.

**Comment by johanneswilm** [comment]
OK, I'll make a document and put them in there, because that's what we decided at the last F2F and then we don't have them floating around github issues, and then we will discuss your point, @annevk, at the F2F next week.

**Comment by johanneswilm** [comment]
The event order draft spec is now in https://github.com/garykac/event-order

**Comment by garykac** [comment]
I am working on a document with an algorithm that specifies the event order in

https://docs.google.com/document/d/1LJQvjEmWZGzVgZnofpvdkxMj1hEnLniD72XD4DLJWx4/edit?usp=sharing

It's still a work-in-progress, but I'm trying to establish the core event ordering and then expand outward. The goal is to have hooks in place so that the Input Event spec retains control over all the pieces that it needs.

**Comment by johanneswilm** [comment]
This generally looks interesting. I noticed it's limited to keyboard based input, while for Input Events we also have to deal with some non-keyboard input. Any suggestion where the events order for those should live?

**Comment by garykac** [comment]
My intent is that by defining the structure for Keyboard/Mouse/Focus/Input/Composition will give us the hooks and entry points needed for other specs to call into so that event order is defined. Ideally, all of the "ordering" information would be in the algorithms in UIEvents so that people could reason about ordering without jumping through multiple specs. The trick is to do that while keeping as much of the Input-specific stuff in the Input Events spec (and similarly for Pointer Events, et al.)

Can you give concrete examples other than:
* custom input devices (that often just generate keyboard or text input)
* speech -> text
* accessibility devices that generate text input

Because scenarios like that are handled (in a hand-wavy way for now) by simply having a defined way to generate the input events for a given text string.

**Comment by johanneswilm** [comment]
How about clicking a menu item in a context menu (for example 'paste')?

**Comment by johanneswilm** [comment]
@garykac See also the examples in the initial post above.

**Comment by patrickhlauke** [comment]
> How about clicking a menu item in a context menu (for example 'paste')?

unless i'm missing something, the actual click/keypress there isn't exposed to the page itself, but it's handled by the browser. the page only sees the contextmenu event being fired initially...what happens inside the context menu is browser-internal.

[ed. ah i probably took that a bit too literally...assume you actually meant in more general terms "when the browser's 'Paste' function is being invoked, in an abstract/general way, which could/should also include keyboard shortcut activation, activation from the application's main menu, etc]

**Comment by garykac** [comment]
> How about clicking a menu item in a context menu (for example 'paste')?

I would expect that to call [handlePaste](https://docs.google.com/document/d/1LJQvjEmWZGzVgZnofpvdkxMj1hEnLniD72XD4DLJWx4/edit#bookmark=id.uo78y0iljsde)() or similar entry points.

I'd like to expose a more general:
* fire beforeinput event of type Xxx with data
* update DOM
* fire input event of type Xxx with data

But I'm not yet sure about the best way to allow general "update DOM" actions. Some sort of callback, but I don't know the best way to express that structure in the spec.

**Comment by garykac** [comment]
> @garykac See also the examples in the initial post above.

AIUI, those are all just examples of the following pattern:

* fire beforeinput (with type and data)
* if not canceled:
   * update DOM
   * fire input (with type and data)

and a generic `fireInputEventsAndMaybeUpdateDOM` routine would handle those cases. That routine could be called from Input Events, or possibly from the Clipboard, History or Drag-n-Drop specs.

**Comment by johanneswilm** [comment]
Right, but there can be other events related to that, such as a paste event, etc. and the order of the beforeinput/input event to those other events needs to be defined somewhere.

**Comment by annevk** [comment]
It seems to me that @garykac's model allows for that. The high-level paste operation would at some point realize it needs to do `fireInputEventsAndMaybeUpdateDOM` and delegates to that.

**Comment by johanneswilm** [comment]
Ok, so in the spec that defines paste it would list the order of events associated with it and somewhere in that list of events it will list `fireInputEventsAndMaybeUpdateDOM`? And where would spell check and history related events be?

The reason this is relevant is that the way JS editors work, they usually can choose from several different events to listen for and so they will often work with backups to work in older browsers or browsers that haven't fully implemented a specific spec, bugs, or just because under circumstances one event gives more information than another one - so if event A was not fired, they will work with event B. But for that to work, they need to be sure that the order of events is the same and that B is not fired before A.

I realize that is not a new issue introduced with @garykac's model, but given that we are touching this area anyway, and this is an issue we have had for a while, it would be good to see if this can be taken care of as well.

**Comment by annevk** [comment]
The order of events follows from how the operations invoke each other (as the operations ultimately dispatch events).

**Comment by johanneswilm** [comment]
@annevk Does that mean that it can be guaranteed that the order of events of these will be the same in all browsers always?

**Comment by annevk** [comment]
I suspect that for certain algorithms we need to allow for platform variation, but in general that would be the idea, yes.

**Comment by garykac** [comment]
The plan is to define a definite order for all of these events as best we can. Reality may restrict us from being able to state that "X always fires before Y", but in that case we would at least document the variation.
